lib/rspec/matchers/expecteds_for_multiple_diffs.rb

```diff
@@ -0,0 +1,118 @@
+module RSpec
+  module Matchers
+    # @api private
+    # Handles list of expected values when there is a need to render
+    # multiple diffs. Also can handle one value.
+    class ExpectedsForMultipleDiffs
+      # @api private
+      # Wraps provided expected value in instance of
+      # ExpectedForMultipleDiffs. If provided value is already an
+      # ExpectedForMultipleDiffs then it just returns it.
+      # @param [Any] expected value to be wrapped
+      # @return [RSpec::Matchers::ExpectedsForMultipleDiffs]
+      def self.from(expected)
+        return expected if self === expected
+        new.add(expected)
+      end
+
+      # @api private
+      # Appends expected value to itself. Returns itself for possible
+      # chaining.
+      # @param [Any] expected value to append to itself
+      # @param [Hash] options optional keyword arguments, for example for_matcher
+      # @return [RSpec::Matchers::ExpectedsForMultipleDiffs]
+      def add(expected, options={})
+        for_matcher = options.fetch(:for_matcher, NoMatcher.new)
+        expected_list.concat(expected_from(expected, ForMatcher.from(for_matcher)))
+        self
+      end
+
+      # @api private
+      # Returns message with diff(s) appended for provided differ
+      # factory and actual value if there are any
+      # @param [String] message original failure message
+      # @param [Proc] differ_factory brand new differ factory method (or proc, block)
+      # @param [Any] actual value
+      # @return [String]
+      def message_with_diff(message, differ_factory, actual)
+        diff = diffs(differ_factory, actual)
+        message = "#{message}\n#{diff}" unless diff.empty?
+        message
+      end
+
+    protected
+
+      # @api private
+      # Returns list of expected values.
+      # @return [Array<RSpec::Matchers::ExpectedsForMultipleDiffs::Expected>]
+      def expected_list
+        @_expected_list ||= []
+      end
+
+    private
+
+      def diffs(differ_factory, actual)
+        expected_list.map do |expected|
+          expected.diffs(differ_factory, actual)
+        end.inject([], :+).join("\n")
+      end
+
+      def expected_from(value, for_matcher)
+        return value.expected_list if self.class === value
```

Comment:
Having methods coerce their input is handy so you don't have to worry about what the argument is (or if an argument has already been coerced), but it feels like defensive programming to have this in a private method.  Given the only callers are in this class, won't it already know what the object is?  It feels like this conversion is here "just in case" and I'd prefer to see the conversions only done where they are truly needed (hopefully in one or maybe two spots?)